### PR TITLE
Fixes #13701 use League\Flysystem for MediaSources

### LINKED ISSUE #13701: Refactor modMediaSource for composer Amazon S3 support
### Summary/Task
Refactor modMediaSource, modFileMediaSource and modS3MediaSource to use [Flysystem](http://flysystem.thephpleague.com/). Flysystem abstracts the S3 to allow MODX to only focus on configuration and implementation. Additionally allow MODX to easily use any of the following: Azure, Copy.com, DigitalOcean Spaces, Dropbox, FTP, GridFS, Memory, Rackspace, ReplicateAdapter, SFTP, WebDAV, PHPCR, ZipArchive or write your own [adapter](http://flysystem.thephpleague.com/creating-an-adapter/).

### Step to reproduce
N/A

### Observed behavior
Legacy AWS S3 code must be updated to use composer. 

### Expected behavior
Keep the modMediaSourceInterface class as close as possible to avoid breaking ExtJS. Potential effected files:

- core/model/modx/processors/browser
- core/model/modx/processors/source 
- core/model/modx/sources - major refactor will happen here, contains the xPDO DB objects and ACL related perms
- core/model/schema/modx.sources.*
- core/model/modx/modcachemanager.class.php
- core/model/modx/processors/system/phpthumb.class.php
- core/model/modx/modtemplatevar.class.php
- manager/controllers/default/system/file/*
- _build/data/transport.core.media_sources.php
- _build/test/Tests/Model/Sources/modMediaSourceTest.php
- _build/test/Tests/Model/Sources/modFileMediaSourceTest.php
- setup/includes/upgrade/common/2.2-media-sources.php
- setup/includes/upgrade/mysql/2.2.0-rc1.php
- setup/includes/upgrade/sqlsrv/2.2.0-rc1.php

### Environment
MODX 3.0+ version

### PR DESCRIPTION
**Pull request by jgulledge19**
### What does it do?
Refactored modMediaSource, modS3MediaSource and modFileMediaSource to use [Flysystem](http://flysystem.thephpleague.com/). In first revision, can now also drag a single file and move between Media sources in the File trees.

### Why is it needed?

1. To be a composer require for AWS S3
2. Abstract the file systems so easily add in others.

### Related issue(s)/PR(s)
#13701

### CODE REVIEW

File: core/model/modx/sources/modmediasource.class.php [hunk 1]
```diff
@@ -326,37 +176,842 @@ public function addError($field,$message) {
...
+    public function getFilesystem()
+    {
+        return $this->filesystem;
+    }
+
+
+    /**
+     * @return FilesystemInterface
+     */
+    public function getAdaptor()
```
[the remark this answers is not in the document]
Comment (jgulledge19) [reply core/model/modx/sources/modmediasource.class.php]: Ah.. yes that is, thanks for catching

File: core/model/modx/sources/modmediasource.class.php [hunk 2]
```diff
@@ -326,37 +176,842 @@ public function addError($field,$message) {
...
+    /**
+     * Return an array of files and folders at this current level in the directory structure
+     *
+     * @param string $path
+     * @return array
+     */
+    public function getContainerList($path)
+    {
+        $properties = $this->getPropertyListWithDefaults();
+        // @TODO can this method be moved elsewhere?  to another class? is it even needed anymore?
```
[the remark this answers is not in the document]
Comment (jgulledge19) [reply core/model/modx/sources/modmediasource.class.php]: I am trying to keep with the existing: modMediaSourceInterface and the args for getContainerList and getObjectsInContainer should be the same and the return array should be the same. I did not want to rework ExtJS to make it all Flysystem. If you are doing custom file stuff, once a source is loaded just do $source->getFilesystem(), that returns the League part.

For any custom source, if they implemented the modMediaSourceInterface it should still work but if you extended modMediaSource then you may need to review.

**Comment by gpsietzema** [comment]
@jgulledge19 We're currently working on MODX3 and think it might be nice to include this. 

We (well, it was @bezumkin really) did some checking on this to see if it breaks anything and it appears to be just fine. Vasily had an own implementation as an extra, https://modx.com/extras/package/swiftmediasourcetype, which only required one change to have it work with yours.

Did you find anything else which breaks? If so what?

Also, what are your thoughts on adding all adapters to MODX by default? It would download a lot of composer dependencies, which isn't something we'd want I guess.

**Comment by gpsietzema** [comment]
It might be an idea to have an extra per adapter. Although I can also imagine having a media sources screen somewhere, where you can simply enable certain adapters. But enabling one of the adapters there could also mean an extra-install as a background-proces. The extra-installation could then do the composer-related actions as well.

**Comment by rthrash** [comment]
I'm with Mark on skipping S3 in the core, too.

**Pull request merged**